qiskit/ignis/measurement/discriminator/iq_discriminators.py

```diff
@@ -496,3 +496,67 @@ def discriminate(self, x_data: List[List[float]]) -> List[str]:
             the discriminated x_data as a list of labels.
         """
         return self._qda.predict(x_data)
+
+
+class SklearnIQDiscriminator(IQDiscriminationFitter):
+    """
+    A generic discriminant analysis discriminator for IQ data that
+    takes an sklearn classifier as an argument.
+    """
+
+    def __init__(self, classifier, cal_results: Union[Result, List[Result]],
+                 qubit_mask: List[int], expected_states: List[str] = None,
```

Comment:
Can you add type hints for classifier? What classifiers do you expect we can use here?

Reply:
sklearn doesn't require classifiers to inherit from any particular base class, so I'm not sure how to proceed (which is why I didn't write anything in the first place).

Reply:
I agree, that's also a valid reason and I don't think we should have a long list of Unions of classifiers as type hints. What we require is that `classifier` has a `fit` method and a `predict` method. I would suggest amending the doc string to reflect this and raise an error if the given classifier does not have these methods. E.g. add to `__init__` the following before doing `self.classifier = classifier`:
```
        for name in ['fit', 'predict']:
            if not callable(getattr(classifier, name, None)):
                raise QiskitError(classifier.__class__.__name__ + 
                                  ' does not have a callable ' + name + 
                                  ' method.')
```
What do you think?
Could you also implement a test for this? I.e. give it a valid classifier and then some dummy variable that does not have `fit` or `predict` and make sure that an error is raised.

Reply:
Looks good! I will add something like this.

For Python 3.8 there is https://www.python.org/dev/peps/pep-0544/, but that's probably too new to rely on right now.

Reply:
Duck type checking added.